pkg/resource/providerconfig.go

```diff
@@ -46,6 +56,119 @@ func IsMissingReference(err error) bool {
 }
 
+// A CredentialExtractor extracts a provider's credentials.
+type CredentialExtractor interface {
+	Extract(context.Context, string, interface{})
+}
+
+// An ExtractionHandlerFn is a function that can extract a specific type of
+// credentials.
+type ExtractionHandlerFn func(ctx context.Context, c interface{}) ([]byte, error)
+
+// EnvLookupFn looks up an environment variable.
+type EnvLookupFn func(string) string
+
+// OsEnvLookup looks up an environment variable from the OS.
+func OsEnvLookup(s string) string {
+	e, _ := os.LookupEnv(s)
+	return e
+}
+
+// ExtractEnvFn extracts credentials from an environment variable.
+func ExtractEnvFn(e EnvLookupFn) ExtractionHandlerFn {
+	return func(ctx context.Context, c interface{}) ([]byte, error) {
+		s, ok := c.(xpv1.CredentialsSelectors)
+		if !ok {
+			return nil, errors.New(errNotCredentialsSelector)
+		}
+		if s.Env == nil {
+			return nil, errors.New(errExtractEnv)
+		}
+		return []byte(e(s.Env.Name)), nil
+	}
+
+}
+
+// ExtractFsFn extracts credentials from the filesystem.
+func ExtractFsFn(fs afero.Fs) ExtractionHandlerFn {
+	return func(ctx context.Context, c interface{}) ([]byte, error) {
+		s, ok := c.(xpv1.CredentialsSelectors)
+		if !ok {
+			return nil, errors.New(errNotCredentialsSelector)
+		}
+		if s.Fs == nil {
+			return nil, errors.New(errExtractFs)
+		}
+		return afero.ReadFile(fs, s.Fs.Path)
+	}
+}
+
+// ExtractSecretFn extracts credentials from a Kubernetes secret.
+func ExtractSecretFn(client client.Client) ExtractionHandlerFn {
+	return func(ctx context.Context, c interface{}) ([]byte, error) {
+		s, ok := c.(xpv1.CredentialsSelectors)
+		if !ok {
+			return nil, errors.New(errNotCredentialsSelector)
+		}
+		if s.SecretRef == nil {
+			return nil, errors.New(errExtractSecretKey)
+		}
+		secret := &corev1.Secret{}
+		if err := client.Get(ctx, types.NamespacedName{Namespace: s.SecretRef.Namespace, Name: s.SecretRef.Name}, secret); err != nil {
+			return nil, errors.Wrap(err, errGetCredentialsSecret)
+		}
+		return secret.Data[s.SecretRef.Key], nil
+	}
+}
+
+// ProviderConfigCredentialExtractor extracts credentials as specified by a ProviderConfig.
+type ProviderConfigCredentialExtractor struct {
+	handlers map[string]ExtractionHandlerFn
+}
+
+// HandlerOverride overrides or adds an extraction handler.
+type HandlerOverride struct {
+	key     string
+	handler ExtractionHandlerFn
+}
+
+// Override adds or overrides an extraction handler.
+func (h *HandlerOverride) Override(handlers map[string]ExtractionHandlerFn) {
+	handlers[h.key] = h.handler
+}
+
+// WithOverride constructs a new HandlerOverride.
+func WithOverride(key string, handler ExtractionHandlerFn) *HandlerOverride {
+	return &HandlerOverride{
+		key:     key,
+		handler: handler,
+	}
+}
+
+// NewProviderConfigCredentialExtractor returns a new credential extractor for ProviderConfigs.
+func NewProviderConfigCredentialExtractor(client client.Client, overrides ...*HandlerOverride) *ProviderConfigCredentialExtractor {
```

Comment:
we pass the client here so that we can have `ExtractSecretFn` be registered by default, but it does feel kind of weird.